File: src/test/scala/us/zoom/data/dfence/providers/snowflake/SnowflakeNonExistentObjectOwnershipTest.scala
package us.zoom.data.dfence.providers.snowflake

import munit.FunSuite
import org.mockito.ArgumentMatchers.{any, anyString}
import org.mockito.Mockito.{doAnswer, mock, reset, when}
import us.zoom.data.dfence.playbook.Playbook
import us.zoom.data.dfence.playbook.model.PlaybookModel
import us.zoom.data.dfence.providers.snowflake.grant.builder.{SnowflakeGrantBuilder, SnowflakeObjectType}
import us.zoom.data.dfence.providers.snowflake.informationschema.SnowflakeObjectsService
import us.zoom.data.dfence.providers.snowflake.models.SnowflakeGrantModel

import java.nio.file.{Files, Path}
import java.util.HashMap
import java.util.concurrent.ForkJoinPool
import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*

/**
 * Tests that revoke statements (GRANT OWNERSHIP ... TO ROLE SECURITYADMIN) are not generated
 * when the object doesn't exist in SnowflakeObjectsService, except when checked by SnowflakeObjectExistsFilter.
 */
class SnowflakeNonExistentObjectOwnershipTest extends FunSuite:

  private val snowflakeGrantsService     = mock(classOf[SnowflakeGrantsService])
  private val snowflakeStatementsService = mock(classOf[SnowflakeStatementsService])
  private val snowflakeObjectsService    = mock(classOf[SnowflakeObjectsService])

  private var snowflakeProvider: SnowflakeProvider = uninitialized

  override def beforeEach(context: BeforeEach): Unit =
    reset(snowflakeGrantsService, snowflakeStatementsService, snowflakeObjectsService)
    snowflakeProvider = new SnowflakeProvider(
      snowflakeStatementsService,
      snowflakeGrantsService,
      snowflakeObjectsService,
      ForkJoinPool.commonPool()
    )

  /**
   * Verifies that revoke statement is not generated when object doesn't exist in SnowflakeObjectsService.
   * The object exists only when checked by SnowflakeObjectExistsFilter (via mock).
   */
  test("should not generate revoke statement for non-existent object ownership") {
    val databaseName  = "DELETE_ME_USER_PROVISIONER_DB"
    val schemaName    = "DELETE_ME_USER_PROVISIONER_TEST_USER1"
    val tableName     = "TABLE_X"
    val fullTableName = s"$databaseName.$schemaName.$tableName"

    // Mock: objectExists returns false by default, true when called from SnowflakeObjectExistsFilter
    doAnswer { invocation =>
      val stackTrace = Thread.currentThread().getStackTrace
      stackTrace.exists(_.getClassName.contains("SnowflakeObjectExistsFilter"))
    }.when(snowflakeObjectsService).objectExists(anyString(), any(classOf[SnowflakeObjectType]))

    val playbookModel = getUserProvisionerPlaybookModel()
    val ownerRoleName = "OWNER_DELETE_ME_USER_PROVISIONER_TEST_USER1"

    val ownershipGrant = new SnowflakeGrantModel(
      "OWNERSHIP",
      "TABLE",
      fullTableName,
      "ROLE",
      ownerRoleName,
      false,
      false,
      false
    )

    val ownershipGrantBuilder = SnowflakeGrantBuilder.fromGrant(ownershipGrant)
    assert(ownershipGrantBuilder != null, "Ownership grant builder should be created")

    val currentGrants = new HashMap[String, SnowflakeGrantBuilder]()
    currentGrants.put(ownershipGrantBuilder.getKey, ownershipGrantBuilder)

    when(snowflakeGrantsService.getGrants(ownerRoleName, false)).thenReturn(currentGrants)

    val ownerRole       = playbookModel.roles().get(ownerRoleName)
    val privilegeGrants = ownerRole.grants()

    val statements = snowflakeProvider.compilePlaybookPrivilegeGrants(
      privilegeGrants,
      ownerRoleName,
      true,
      true,
      false,
      playbookModel,
      false,
      ownerRole.unsupportedRevokeBehavior()
    )

    val expectedRevokeStatement =
      s"""GRANT OWNERSHIP ON TABLE "$databaseName"."$schemaName"."$tableName" TO ROLE SECURITYADMIN COPY CURRENT GRANTS;"""
    val allStatements           = statements.asScala.flatMap(_.asScala).toList

    assert(
      !allStatements.contains(expectedRevokeStatement),
      "Revoke statement should be filtered out by SnowflakeOwnedObjectFilter when another role has ownership"
    )
  }

  private def getUserProvisionerPlaybookModel(): PlaybookModel =
    val resourceUrl =
      getClass.getClassLoader.getResource("test-data/snowflake-non-existent-object-ownership-test.yml")
    if resourceUrl == null then
      throw new RuntimeException(
        "Could not find snowflake-non-existent-object-ownership-test.yml in test resources"
      )
    val yamlString  = Files.readString(Path.of(resourceUrl.toURI))
    Playbook.parse(yamlString, new HashMap())

File: src/test/java/us/zoom/data/dfence/providers/snowflake/revoke/SnowflakeNonExistentObjectOwnershipTest.java
package us.zoom.data.dfence.providers.snowflake.revoke;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ForkJoinPool;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import us.zoom.data.dfence.playbook.Playbook;
import us.zoom.data.dfence.playbook.model.PlaybookModel;
import us.zoom.data.dfence.providers.snowflake.SnowflakeGrantsService;
import us.zoom.data.dfence.providers.snowflake.SnowflakeProvider;
import us.zoom.data.dfence.providers.snowflake.SnowflakeStatementsService;
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeGrantBuilder;
import us.zoom.data.dfence.providers.snowflake.informationschema.SnowflakeObjectsService;
import us.zoom.data.dfence.providers.snowflake.models.PartitionedGrantStatements;
import us.zoom.data.dfence.providers.snowflake.models.SnowflakeGrantModel;

/**
 * Tests that revoke statements (GRANT OWNERSHIP ... TO ROLE SECURITYADMIN) are not generated when
 * the object doesn't exist in SnowflakeObjectsService, except when checked by
 * SnowflakeObjectExistsFilter.
 */
class SnowflakeNonExistentObjectOwnershipTest {

  @Mock private SnowflakeGrantsService snowflakeGrantsService;

  @Mock private SnowflakeStatementsService snowflakeStatementsService;

  @Mock private SnowflakeObjectsService snowflakeObjectsService;

  private SnowflakeProvider snowflakeProvider;

  @BeforeEach
  void setUp() {
    MockitoAnnotations.openMocks(this);
    snowflakeProvider =
        new SnowflakeProvider(
            snowflakeStatementsService,
            snowflakeGrantsService,
            snowflakeObjectsService,
            ForkJoinPool.commonPool());
  }

  /**
   * Verifies that revoke statement is not generated when object doesn't exist in
   * SnowflakeObjectsService. The object exists only when checked by SnowflakeObjectExistsFilter
   * (via mock).
   */
  @Test
  void shouldNotGenerateRevokeStatementForNonExistentObjectOwnership()
      throws IOException, URISyntaxException {
    String databaseName = "DELETE_ME_USER_PROVISIONER_DB";
    String schemaName = "DELETE_ME_USER_PROVISIONER_TEST_USER1";
    String tableName = "TABLE_X";
    String fullTableName = databaseName + "." + schemaName + "." + tableName;

    // Mock: objectExists returns false by default, true when called from
    // SnowflakeObjectExistsFilter
    doAnswer(
            invocation -> {
              StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
              for (StackTraceElement element : stackTrace) {
                if (element.getClassName().contains("SnowflakeObjectExistsFilter")) {
                  return true;
                }
              }
              return false;
            })
        .when(snowflakeObjectsService)
        .objectExists(anyString(), any());

    PlaybookModel playbookModel = getUserProvisionerPlaybookModel();
    String ownerRoleName = "OWNER_DELETE_ME_USER_PROVISIONER_TEST_USER1";

    SnowflakeGrantModel ownershipGrant =
        new SnowflakeGrantModel(
            "OWNERSHIP", "TABLE", fullTableName, "ROLE", ownerRoleName, false, false, false);

    SnowflakeGrantBuilder ownershipGrantBuilder = SnowflakeGrantBuilder.fromGrant(ownershipGrant);
    assertNotNull(ownershipGrantBuilder, "Ownership grant builder should be created");

    HashMap<String, SnowflakeGrantBuilder> currentGrants = new HashMap<>();
    currentGrants.put(ownershipGrantBuilder.getKey(), ownershipGrantBuilder);

    when(snowflakeGrantsService.getGrants(ownerRoleName, false)).thenReturn(currentGrants);

    var ownerRole = playbookModel.roles().get(ownerRoleName);
    var privilegeGrants = ownerRole.grants();

    PartitionedGrantStatements partitionedStatements =
        snowflakeProvider.compilePlaybookPrivilegeGrants(
            privilegeGrants,
            ownerRoleName,
            true,
            true,
            false,
            playbookModel,
            false,
            ownerRole.unsupportedRevokeBehavior());
    List<List<String>> statements = new ArrayList<>();
    statements.addAll(partitionedStatements.ownershipStatements());
    statements.addAll(partitionedStatements.nonOwnershipStatements());

    String expectedRevokeStatement =
        String.format(
            "GRANT OWNERSHIP ON TABLE \"%s\".\"%s\".\"%s\" TO ROLE SECURITYADMIN COPY CURRENT GRANTS;",
            databaseName, schemaName, tableName);
    List<String> allStatements = statements.stream().flatMap(List::stream).toList();

    assertFalse(
        allStatements.contains(expectedRevokeStatement),
        "Revoke statement should be filtered out by SnowflakeOwnedObjectFilter when another role has ownership");
  }

  private PlaybookModel getUserProvisionerPlaybookModel() throws IOException, URISyntaxException {
    var resourceUrl =
        getClass()
            .getClassLoader()
            .getResource("test-data/snowflake-non-existent-object-ownership-test.yml");
    if (resourceUrl == null) {
      throw new RuntimeException(
          "Could not find snowflake-non-existent-object-ownership-test.yml in test resources");
    }
    String yamlString = Files.readString(Path.of(resourceUrl.toURI()));
    return Playbook.parse(yamlString, new HashMap<>());
  }
}
